src/nbl/video/IDescriptorPool.cpp

```diff
@@ -156,14 +169,21 @@ void IDescriptorPool::deleteSetStorage(IGPUDescriptorSet* set)
     const auto mutableSamplerCount = set->getLayout()->getTotalMutableSamplerCount();
     if (mutableSamplerCount > 0)
     {
-        const uint32_t allocatedOffset = set->getMutableSamplerStorageOffset();
+        const uint32_t allocatedOffset = set->m_storageOffsets.getMutableSamplerOffset();
         assert(allocatedOffset != ~0u);
 
         std::destroy_n(getMutableSamplerStorage() + allocatedOffset, mutableSamplerCount);
 
         if (allowsFreeing())
             m_descriptorAllocators[static_cast<uint32_t>(asset::IDescriptor::E_TYPE::ET_COUNT)]->free(allocatedOffset, mutableSamplerCount);
-    } 
+    }
+
+    m_descriptorSetAllocator.free_addr(set->m_storageOffsets.getSetOffset(), 1);
+    if (m_descriptorSetAllocator.get_allocated_size() == 0)
+        reset();
```

Comment:
> otherwise the reset will attempt to call deleteSetStorage on it

Are you sure? Because just above I free the DS from the allocator with:
```C++
m_descriptorSetAllocator.free_addr(set->m_storageOffsets.getSetOffset(), 1);
```
then `reset` uses the `IteratablePoolAddressAllocator`'s iterator to iterate only on the non-freed addresses (judging from [this](https://github.com/Devsh-Graphics-Programming/Nabla/blob/2275fb0ff65169901caedd61e9fdb86e19214f73/include/nbl/core/alloc/IteratablePoolAddressAllocator.h#L99) and [this](https://github.com/Devsh-Graphics-Programming/Nabla/pull/345#discussion_r1116470839)).

Reply:
also implicit `reset` is only beneficial when `!allowsFreeing()` because a individually free-able pools allocators go back to initial state after all DS have been freed

Reply:
> then reset uses the IteratablePoolAddressAllocator's iterator to iterate only on the non-freed addresses (judging from [this](https://github.com/Devsh-Graphics-Programming/Nabla/blob/2275fb0ff65169901caedd61e9fdb86e19214f73/include/nbl/core/alloc/IteratablePoolAddressAllocator.h#L99) and https://github.com/Devsh-Graphics-Programming/Nabla/pull/345#discussion_r1116470839).

ok I guess that prevents us from infinite recursion and double frees.

Still best to move this to the end as keeping it here has no benefits (tail recursion and stuff).

Also no point calling `reset` if `allowsFreeing()`

Reply:
> Still best to move this to the end as keeping it here has no benefits (tail recursion and stuff).

I think it won't be possible because after you `set = nullptr` you obviously won't be able to do `set->m_pool = nullptr` and `set->m_pool = nullptr` must be done after calling `reset` uses the `m_pool`'s member `m_descriptorSetAllocator`,

Reply:
> Also no point calling reset if allowsFreeing()

If we do this https://github.com/Devsh-Graphics-Programming/Nabla/pull/345#discussion_r1122857714 I think we would have to call `reset` regardless to set nullify the pointers in `m_allocatedDescriptorSets`.